### DISCUSSION TITLE
Enabling and Using the DEN Service

### DISCUSSION THREAD

**Discussion by HObaid00** [riebl/artery#350] [Q&A] [answered]
Hello everyone, 
I'm trying to build a traffic network that uses the DEN service in Artery for cooperative purposes in traffic. My problem specifically is that I don't get any response from the DEN service when implementing it into my setup. This leads me to ask if anyone has experience calling the service and would like to show an example of how they set it up for it. 

Thank you in advance,
Best regards.

**Comment by riebl** [comment, marked as answer]
Hi @3gles00,

the DEN service itself is rather generic and specific DEN messages are generated whenever a particular use case is fulfilled.
For example, https://github.com/riebl/artery/blob/master/scenarios/storyboard/den-use-cases.xml enables three use cases.
The use cases check their specific triggering conditions periodically and they are also filling a blank DENM accordingly. I think the `EmergencyBrakeLight` use case class is a good and not too complicated example.
Of course, you will not observe any DENMs on the channel if the triggering conditions are never fulfilled in your setup.

Best regards,
Raphael

**Reply by HObaid00** [reply]
Hi @riebl, thank you for the examples.

The corresponding  .ned file is the criteria for sending the DENM. I see "indicate" functions for each corresponding UseCase and in the UseCase.h file says that these "indicate" functions will be reacted to immediately; if I would like to respond to the incoming DENM, is it here that it needs to be edited? And are the DENM case sensitive, or do they all go through the UseCase.cc file?

Thank you in advance,
Best regards.

**Reply by riebl** [reply]
The moment a `DenService` receives a DEN message, it invokes the `indicate` methods of all its use cases. The use cases can freely decide how to deal with this message, e.g. ignore it or send a response such as `ImpactReductionUseCase` immediately.
